### PR TITLE
Plan Cache: Replace the cache implementation

### PR DESCRIPTION
**Pull request by vmg**
Hiii everyone! First PR to Vitess in a while. Let's get into it!

## Description

The cache for planned queries, which exists both in the `vtgates` and the `vttablets`, has not been behaving well. We have had to disable these caches altogether because they were causing OOM errors. We believe the main cause triggering these errors are **bulk insert queries**, which come in different sizes for statement counts and hence will occupy too much cache space even after normalization before caching.

After some pairing with @sougou, we have considered the possibility of disabling caching altogether for `INSERT` statements, since these are the ones known to cause particular problems. Before doing this, I promised to gather some numbers on how expensive are the query plans for these statements to prepare. The results are not encouraging: on average, `INSERT` statements are just as expensive as any other statement to plan (as seen on the commit message for https://github.com/vitessio/vitess/commit/fe240ec19260d4769d04d176e220047bce1282ee), so disabling their caching is going to cause a performance regression.

## Next Steps

Despite the fact that we've tracked down the OOM issues to batch-`INSERT` queries, I think the underlying issue is clear upon reviewing the implementation of our cache: the main issue we're facing (which batch inserts definitely exacerbate) is that our plan cache is too primitive for our use case. Most notably, **it does not have an admission policy**. Caches for database systems have historically always always had an admission policy, whose goal is preventing extreme corner cases from taking over the cache. In our case, this is batch inserts queries; an equivalent in a traditional relational database system would be a full-table scan, which would page into cache a lot of pages in the database that will only be read once. These kind of pathological access patterns cause **cache pollution** by bringing into cache a lot of data that is never going to be read again. 

My belief is that we are going to see a significant performance improvement by replacing the `LRUCache` implementation in Vitess with a LFU w/ eviction policy that cannot be trivially polluted. I think would be a minimal effort which I'd like to undertake next week, and it should fix both the cache memory growth issues we're seeing **and improve cache performance overall** since the current implementation with a Map + Linked List uses a single global lock for the whole cache, which right now is actively being contended by all the query goroutines in a `vtgate`. Obviously, it would also allow us to enable `INSERT` plan caching again, since batch inserts ought to be a corner case that will no longer pollute our cache. This is something which I intend to verify in an integration test.

@sougou @enisoc I would like your feedback on this before I get started on Monday.

## Checklist
- [ ] Should this PR be backported?
- [ ] Tests were added or are not required
- [ ] Documentation was added or is not required

## Impacted Areas in Vitess
Components that this PR will affect:

- [x]  Query Serving
- [ ]  VReplication
- [ ]  Cluster Management
- [ ]  Build 
- [ ]  VTAdmin

### CODE REVIEW

File: go/vt/vtgate/executor.go [hunk 1]
```diff
@@ -1295,11 +1295,6 @@ func (e *Executor) getPlan(vcursor *vcursorImpl, sql string, comments sqlparser.
 	ignoreMaxMemoryRows := sqlparser.IgnoreMaxMaxMemoryRowsDirective(stmt)
 	vcursor.SetIgnoreMaxMemoryRows(ignoreMaxMemoryRows)
 
-	planKey := vcursor.planPrefixKey() + ":" + sql
```
Comment (systay) [inline go/vt/vtgate/executor.go:1298]: Why are you removing this?

**Comment by systay** [comment]
Nice stuff! I like your thinking, and you are probably 100% correct.

**Comment by vmg** [comment]
@systay I've started working on a new cache implementation. The first step has been reducing the API surface for the existing cache and abstracting it behind an `interface` so it can be hotswapped with a different one. I've marked the PR as a draft while I look at this. Would you mind getting started with the review of these changes? I've split them into individual commits that should be easier to review. Cheers!

File: go/vt/vtgate/planbuilder/plan_test.go [hunk 2]
```diff
@@ -617,6 +618,45 @@ func BenchmarkPlanner(b *testing.B) {
...
+		v:             loadSchema(b, "schema_test.json"),
+		sysVarEnabled: true,
+		version:       V4,
+	}
+
+	var dmlCases []testCase
+	var selectCases []testCase
+
+	for tc := range iterateExecFile("dml_cases.txt") {
+		if tc.output2ndPlanner != "" {
```
Comment (systay) [inline go/vt/vtgate/planbuilder/plan_test.go:632]: `output2ndPlanner` is an artifact of this project: #7280

it's the plan generated by the new gen4 planner. for this work, you should just ignore it and add all queries.

File: go/cache/cache.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
+package cache
```
Comment (systay) [inline go/cache/cache.go:1]: License headers

File: go/cache/lru_cache_test.go [hunk 4]
```diff
@@ -26,24 +26,24 @@ type CacheValue struct {
...
 	return cv.size
 }
 
 func TestInitialState(t *testing.T) {
 	cache := NewLRUCache(5)
-	l, sz, c, e, _ := cache.Stats()
-	if l != 0 {
-		t.Errorf("length = %v, want 0", l)
+	stats := cache.Stats()
+	if stats.Length != 0 {
```
Comment (systay) [inline go/cache/lru_cache_test.go:36]: nit: we try to move to testify whenever we have to change these types of assertions in the code

Comment (vmg) [reply go/cache/lru_cache_test.go:36]: SGTM. I've noticed that we're mixing testify and classic assertions. I'll update the tests. :+1:

**Comment by vmg** [comment]
Can I assume that the failure in the race detector is a flake? It doesn't look related to any of my coder changes.

**Comment by systay** [comment]
I think the reason the executor tries twice is to save on parsing, normalization and AST rewriting. I believe last we checked these where non-trivial parts of the full process. Don't trust my memory though - we measure to see if it is worth double checking or not.

**Comment by systay** [comment]
I _really_ like your new cleaner API.

**Comment by systay** [comment]
> Can I assume that the failure in the race detector in CI is a flake? It doesn't look related to any of my code changes.

#6067 it's a know flaky test. I've restarted it. 

`TestQueryStats` is a real failure though, right?

**Comment by vmg** [comment]
I've just pushed the next two iterative steps that are going to be required for a proper cache implementation.

My work for today has consisted on fixing the most glaring shortcoming in the current cache: the fact that the size of the cache is measured in **entries** as opposed to **bytes**. Right now, both `vtgate` and `vttablet` have configuration options that let the user choose how many entries can a cache hold. This is not a good approach for production systems, because it means that **the memory consumption of the cache cannot be tuned _a priori_**, as the only configurable limit is the number of entries, which have wildly different sizes between them. I believe it's crucial for Vitess to be able to finely tune its cache behaviour in production systems based on the amount of memory available on a host.

Hence, commit d7c4172 introduces two new configuration settings for `vtgate` and `vttablet` that allow you to tune the exact memory size of the cache instead of the number of entries. The old configuration options for number of entries have been left as-is but marked as deprecated. Vitess now knows how to convert the old entry-count configuration, when explicitly provided by the user, into a rough memory limit based on the average size of a cache entry; this provides backwards compatibility until we're eventually ready to remove the old config options.

Lastly, in order to make the new cache limits work in practice, I've had to implement a rough calculation of memory consumption for all `Plan` objects. That is proving itself to be a pain! The implementation I'm proposing here is more-or-less accurate by manually recursing the data structures and adding up their consumption; I must say I'm not in love with it. Another possible approach would be using runtime reflection to measure the size (something which would require a lot less code but may potentially be slower). I intend to attempt such implementation tomorrow. I would love to hear your feedback on other possible ways of implementing this functionality that aren't overly verbose.

My remaining work for today is going to be getting the tests green. :green_apple:

File: go/cache/null.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+// nullCache is a no-op cache that does not store items
+type nullCache struct{}
+
+func (n *nullCache) Get(_ string) (interface{}, bool) {
```
Comment (systay) [inline go/cache/null.go:22]: nit: didn't the linter complain about comments?

File: go/vt/vtgate/vtgate.go [hunk 6]
```diff
@@ -178,8 +180,14 @@ func Init(ctx context.Context, serv srvtopo.Server, cell string, tabletTypesToWa
 	resolver := NewResolver(srvResolver, serv, cell, sc)
 	vsm := newVStreamManager(srvResolver, serv, cell)
 
+	// If the legacy queryPlanCacheSize is set, override the value of the new queryPlanCacheSizeBytes
+	// approximating the total size of the cache with the average size of an entry
+	if *queryPlanCacheSize != 0 {
+		*queryPlanCacheSizeBytes = *queryPlanCacheSize * engine.AveragePlanSize
```
Comment (systay) [inline go/vt/vtgate/vtgate.go:186]: should this lead to us logging a warning about a deprecated flag being used?

File: go/vt/sqlparser/parsed_query.go [hunk 7]
```diff
@@ -44,6 +45,16 @@ func NewParsedQuery(node SQLNode) *ParsedQuery {
 	return buf.ParsedQuery()
 }
 
+// CachedSize returns the approximate size that this query takes when cached in memory
+func (pq *ParsedQuery) CachedSize() int64 {
+	if pq == nil {
+		return 0
+	}
+	return int64(unsafe.Sizeof(ParsedQuery{})) +
```
Comment (systay) [inline go/vt/sqlparser/parsed_query.go:53]: historically, we have relegated the use of unsafe to https://github.com/vitessio/vitess/blob/master/go/hack/hack.go. Is it time to let go of that rule, @sougou? (I think it is)

**Comment by systay** [comment]
the commit `cache: configure using total memory usage` is missing the sign-off thingie

**Comment by vmg** [comment]
@systay I keep seeing two test failures that look very unrelated to my changes. Am I wrong here? Are the tests usually this flaky?

**Comment by shlomi-noach** [comment]
@vmg we do suffer from flakiness and are taking flaky tests down as time permits... It's unfortunately not uncommon .

**Comment by vmg** [comment]
Replaced with https://github.com/vitessio/vitess/pull/7439
